aesara/scalar/math.py

```diff
@@ -1493,3 +1495,174 @@ def c_code(self, *args, **kwargs):
+            Derivative of hyp2f1 wrt a
+
+            """
+
+            if abs(z) >= 1:
+                raise NotImplementedError("Gradient not supported for |z| >= 1")
+
+            else:
+                term1 = _infinisum(
+                    lambda k: (
+                        (gamma(a + k) / gamma(a))
+                        * (gamma(b + k) / gamma(b))
+                        * psi(a + k)
+                        * (z**k)
+                    )
+                    / (gamma(c + k) / gamma(c))
+                    * gamma(k + 1)
+                )
+                term2 = psi(a) * hyp2f1(a, b, c, z)
+
+                return term1 - term2
+
+        def _hyp2f1_db(a, b, c, z):
+            """
+            Derivative of hyp2f1 wrt b
+            """
+
+            if abs(z) >= 1:
+                raise NotImplementedError("Gradient not supported for |z| >= 1")
+
+            else:
+                term1 = _infinisum(
+                    lambda k: (
+                        (gamma(a + k) / gamma(a))
+                        * (gamma(b + k) / gamma(b))
+                        * psi(b + k)
+                        * (z**k)
+                    )
+                    / (gamma(c + k) / gamma(c))
+                    * gamma(k + 1)
+                )
+                term2 = psi(b) * hyp2f1(a, b, c, z)
+
+                return term1 - term2
+
+        def _hyp2f1_dc(a, b, c, z):
+            """
+            Derivative of hyp2f1 wrt c
+            """
+            if abs(z) >= 1:
+                raise NotImplementedError("Gradient not supported for |z| >= 1")
+
+            else:
+                term1 = psi(c) * hyp2f1(a, b, c, z)
+                term2 = _infinisum(
+                    lambda k: (
+                        (gamma(a + k) / gamma(a))
+                        * (gamma(b + k) / gamma(b))
+                        * psi(c + k)
+                        * (z**k)
+                    )
+                    / (gamma(c + k) / gamma(c))
+                    * gamma(k + 1)
+                )
+                return term1 - term2
+
+        def _hyp2f1_dz(a, b, c, z):
+            """
+            Derivative of hyp2f1 wrt z
+            """
+
+            return ((a * b) / c) * hyp2f1(a + 1, b + 1, c + 1, z)
+
+        if wrt == 0:
+            return _hyp2f1_da(a, b, c, z)
+        elif wrt == 1:
+            return _hyp2f1_db(a, b, c, z)
+        elif wrt == 2:
+            return _hyp2f1_dc(a, b, c, z)
+        elif wrt == 3:
+            return _hyp2f1_dz(a, b, c, z)
+
+    def c_code(self, *args, **kwargs):
+        raise NotImplementedError()
+
+
+hyp2f1_der = Hyp2F1Der(upgrade_to_float, name="hyp2f1_der")
+
+
+def poch(z: ScalarType, m: ScalarType) -> ScalarVariable:
+    """
+    Pochhammer symbol (rising factorial) function.
+
+    """
+    return gamma(z + m) / gamma(z)
+
+
+def factorial(n: ScalarType) -> ScalarVariable:
+    """
+    Factorial function of a scalar or array of numbers.
+
+    """
+    return gamma(n + 1)
```

Comment:
These need to be in `aesara.tensor.special` and `aesara.tensor.gamma` needs to be used, because that `gamma` is an `Elemwise` `Op`.  That `Elemwise` version of `gamma` is created from the scalar `aesara.scalar.math.Gamma` `Op` via the `scalar_elemwise` decorator in `aesara.tensor.math`, which creates the `Elemwise` like `gamma = Elemwise(Gamma, ...)`.

In general, `Elemwise` `Op`s are the most common way that `ScalarOp`s are used in Aesara, since they add broadcasting to scalar operations.